### PR TITLE
Abstract operator

### PR DESCRIPTION
**Pull request by tombentley**
### Type of change

- Refactoring

### Description

This PR is a refactoring motivated in part by @pamelipluas's work on #1468. See the comments there around resource deletion to understand the motivation what I'm doing here.

* In this PR I factor out an `Operator` interface, an `AbstractOperator` partial implementation of it. 
* `KafkaUserOperator` and `AbstractAssemblyOperator` become subclasses of `AbstractOperator`. 
* The connector operator will become another subclass of `AbstractOperator` in due course.
* TBH I'm not wedded to having an `Operator` interface. Perhaps only the `AbstractOperator` is necessary, but the interface itself doesn't depend on fabric8 and not every operator needs to do things in the way `AbstractOperator` does.
* I ended up using `listAsync()` rather than `list()` in the implementation of `reconcileAll()`.
* A minor point: Previously the KUO and AAO used different timeouts on the lock which prevents concurrent reconciliation of a given resource. They now use the a single timeout (which is shorter than the one used for the AAO previously.
* I've also removed the pointless `enum ResourceType` (a `String kind` serves just as well).

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

**Comment by tombentley** [comment]
@strimzi-ci run tests

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 1]
```diff
@@ -2723,4 +2725,24 @@ String getInternalServiceHostname(String serviceName)    {
     private Date dateSupplier() {
         return new Date();
     }
+
+    /**
+     * @param current Previous ConfigMap
```
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: The parameter is current but the description says Previous, it sounds confusing :)

File: operator-common/src/main/java/io/strimzi/operator/common/OperatorWatcher.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.Watcher;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.function.Consumer;
+
+public class OperatorWatcher<T extends HasMetadata> implements Watcher<T> {
```
Comment (ppatierno) [inline operator-common/src/main/java/io/strimzi/operator/common/OperatorWatcher.java]: A little bit of javadoc?

**Comment by tombentley** [comment]
@strimzi-ci run tests testcase=io.strimzi.systemtest.RecoveryST#testRecoveryFromZookeeperStatefulSetDeletion

**Pull request merged**